src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineHelper.cs

```diff
@@ -0,0 +1,365 @@
+        /// </remarks>
+        public static int CompareSymbolOrder(DocumentSymbolViewModel x, DocumentSymbolViewModel y, ITextSnapshot oldSnapshot, ITextSnapshot currentSnapshot)
+        {
+            var xStartPosition = oldSnapshot.GetLineFromLineNumber(x.StartPosition.Line).Start.Position + x.StartPosition.Character;
+            var yStartPosition = oldSnapshot.GetLineFromLineNumber(y.StartPosition.Line).Start.Position + y.StartPosition.Character;
+
+            var xCurrentStartPosition = new SnapshotPoint(currentSnapshot, xStartPosition).Position;
+            var yCurrentStartPosition = new SnapshotPoint(currentSnapshot, yStartPosition).Position;
+
+            return xCurrentStartPosition - yCurrentStartPosition;
+        }
+
+        /// <summary>
+        /// Compares the type (SymbolKind) of two DocumentSymbolViewModels.
+        /// </summary>
+        public static int CompareSymbolType(DocumentSymbolViewModel x, DocumentSymbolViewModel y)
+        {
+            if (x.SymbolKind == y.SymbolKind)
+                return x.Name.CompareTo(y.Name);
+
+            return x.SymbolKind - y.SymbolKind;
+        }
+
+        /// <summary>
+        /// Sorts and returns an immutable array of DocumentSymbolViewModels based on a SortOption.
+        /// </summary>
+        /// <remarks>
+        /// The parameter oldSnapshot refers to the snapshot used when the LSP document symbol request was made to 
+        /// obtain the symbol and currentSnapshot refers to the latest snapshot in the editor. These parameters are
+        /// required to obtain the latest DocumentSymbolViewModel range positions in the new snapshot (to sort by order).
+        /// </remarks>
+        public static ImmutableArray<DocumentSymbolViewModel> Sort(
+            ImmutableArray<DocumentSymbolViewModel> documentSymbolModels,
+            SortOption sortOption,
+            ITextSnapshot? oldSnapshot,
+            ITextSnapshot? currentSnapshot)
+        {
+            // We want to log which sort option is used
+            Logger.Log(sortOption switch
+            {
+                SortOption.Name => FunctionId.DocumentOutline_SortByName,
+                SortOption.Order => FunctionId.DocumentOutline_SortByOrder,
+                SortOption.Type => FunctionId.DocumentOutline_SortByType,
+                _ => throw new NotImplementedException(),
+            });
+
+            // Sort the top-level DocumentSymbolViewModels
+            var sortedDocumentSymbolModels = sortOption switch
+            {
+                SortOption.Name => documentSymbolModels.Sort((x, y) => x.Name.CompareTo(y.Name)),
+                SortOption.Order => documentSymbolModels.Sort((x, y) =>
+                {
+                    if (oldSnapshot is null)
+                        throw new ArgumentNullException(nameof(oldSnapshot));
+                    else if (currentSnapshot is null)
+                        throw new ArgumentNullException(nameof(currentSnapshot));
```

Comment:
Should something have prevented this from happening earlier? Seems like this is asking for trouble if it only throws in some branches?

Reply:
Outdated - these fields are no longer needed